### PR TITLE
Export image

### PR DESCRIPTION
**Pull request by DierF**
Refactor the process of drawing diagram.
Use pygame and PIL for drawing image.
Export command in CLI and export option is GUI.

### CODE REVIEW

File: src/Models/uml_image.py [hunk 1]
```diff
@@ -0,0 +1,270 @@
...
+                curr += 1
+            pygame.draw.line(framebuffer, (0, 0, 0), (cls_x + self.margin, cls_y + curr * self.line_height + self.margin),
+                            (cls_x + cls_width + self.margin, cls_y + curr * self.line_height + self.margin), 3)
+            for text_method in text_methods:
+                text_surface = font.render(text_method, True, (0, 0, 0))
+                framebuffer.blit(text_surface, (cls_x + (cls_width - len(text_method) * self.letter_width) // 2 + self.margin,
+                                                cls_y + curr * self.line_height + self.margin))
+                curr += 1
+
+    def __vec(self, p1: list[int], p2: list[int]):
```
Comment (pwfreedm) [inline src/Models/uml_image.py:212]: definitely could've given all these operations to a struct as well.

**Review by pwfreedm — CHANGES_REQUESTED** [review]
I don't know that any of these changes are reasonable to expect between now and tomorrow night. Given that everything appears logical (if poorly organized), just look over the comments, if you think any of the changes make sense to attempt now then go for it, otherwise resolve them and I'll accept once they're done.

File: src/Controllers/controller.py [hunk 2]
```diff
@@ -45,13 +49,14 @@ def run(self):
...
-                # since the Diagram is not changed if Exception is thrown in most cases
-                # For example: if cmd contains invalid arguments, the diagram is not touched at all,
-                #               but this undo will undo to last state, which is not expected.
-                # self._states.undo()
                 continue
-            self._controller.draw(self._diagram)
+            if isinstance(self._controller, GUI_Controller):
+                camera_pos = self._controller.get_camera_pos()
+                viewport_size = self._controller.get_viewport_size()
+                framebuffer, class_boxes = self._image.draw_framebuffer(self._diagram, camera_pos)
```
[the remark this answers is not in the document]
Comment (DierF) [reply src/Controllers/controller.py:56]: camera_pos is used to find the right position of image to draw. If we know where the camera is and the size of the viewport, we only need to draw a small image that fits the camera's view.

File: src/Models/uml_image.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+        font = pygame.font.Font(None, self.font_size)
+        # draw relationship arrows
+        self.__draw_relationship_arrows(framebuffer, diagram, class_rects)
+        # draw class boxes
+        self.__draw_class_boxes(framebuffer, font, class_rects)
+        return framebuffer
+
+    # helpers #
+
+    def __generate_class_boxes_and_class_rects_and_boarders(self, diagram: UML_Diagram):
```
[the remark this answers is not in the document]
Comment (DierF) [reply src/Models/uml_image.py:49]: I know this looks a mess, I just don't have time to organize them better.

File: src/Models/uml_image.py [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+        top_border = 0
+        bot_border = 0
+        class_boxes = []
+        class_rects = []
+        for cls in diagram.get_all_classes():
+            text_cls_width = len(cls.get_name()) * self.letter_width
+            text_cls_height = self.line_height
+            text_fields = []
+            for field in cls.get_fields():
+                text_field = '{} {}'.format(field.get_type(), field.get_name())
```
[the remark this answers is not in the document]
Comment (DierF) [reply src/Models/uml_image.py:61]: list visitor is in Views? This format is the easiest to do, we could change it later if needed.

File: src/Models/uml_image.py [hunk 5]
```diff
@@ -0,0 +1,270 @@
...
+        self.letter_width = 10
+        self.margin = 500
+        self.background_color = (100, 100, 100)
+        self.font_size = 25
+        self._viewport_width = 1000
+        self._viewport_height = 800
+        self._framebuffer = pygame.Surface((self._viewport_width, self._viewport_height))
+
+    def draw_framebuffer(self, diagram: UML_Diagram, camera_pos: tuple[int, int]):
+        class_boxes, class_rects, _, _ = self.__generate_class_boxes_and_class_rects_and_boarders(diagram)
```
[the remark this answers is not in the document]
Comment (DierF) [reply src/Models/uml_image.py:21]: All these three things are generated in one loop, separate them will likely to make performance worse. Make a class to hold those data sounds like a good idea. We can do that if we have time to optimze the code.

File: src/Controllers/controller.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (pwfreedm) [reply src/Controllers/controller.py:56]: ah, I was referring to framebuffer and class_boxes. Apologies for the lack of clarity.

File: src/Models/uml_image.py [hunk 6]
```diff
@@ -0,0 +1,270 @@
...
+                else:
+                    pygame.draw.arc(framebuffer, (0, 0, 0), draw_area, 0, 2 * math.pi, 5)
+                    # draw arrow
+                    if rel.get_type() == 'Aggregation':
+                        self.__draw_diamond(framebuffer, [src_x - radius, src_y + radius - 5], [src_x, src_y + radius - 5], (255, 255, 255))
+                    elif rel.get_type() == 'Composition':
+                        self.__draw_diamond(framebuffer, [src_x - radius, src_y + radius - 5], [src_x, src_y + radius - 5], (0, 0, 0))
+                    elif rel.get_type() == 'Inheritance':
+                        self.__draw_triangle(framebuffer, [src_x - radius, src_y + radius - 5], [src_x, src_y + radius - 5], (255, 255, 255))
+                continue
```
[the remark this answers is not in the document]
Comment (DierF) [reply src/Models/uml_image.py:136]: This is to skip to the next iteration(next arrow) of the first for loop.

**Review by pwfreedm — APPROVED** [review]

File: src/Models/uml_image.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (DierF) [reply src/Models/uml_image.py:136]: Didn't want to make the following code in a else block with one more level of indentation.

**Pull request merged**